### PR TITLE
support ks3 objstore

### PR DESCRIPTION
**Pull request by lizw91**
## About ks3

The website links：https://www.ksyun.com/nv/product/KS3.html

## Changes

- add ks3 object driver

## Verification

- e2eTest pass

### CODE REVIEW

File: pkg/objstore/ks3/ks3.go [hunk 1]
```diff
@@ -0,0 +1,521 @@
...
+		}
+	}
+	// partSize cannot be larger than object size.
+	partSize := b.partSize
+	if size < int64(partSize) {
+		err := b.putFile(name, r)
+		if err != nil {
+			return err
+		}
+	} else {
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:148]: Return early will be better :)

File: pkg/objstore/ks3/ks3.go [hunk 2]
```diff
@@ -0,0 +1,521 @@
...
+		})
+		if compErr != nil {
+			return fmt.Errorf("bucket:%s,objectKey:%s,CompleteMultipartUpload failed,out:%v,err:%v", b.name, b.name, compRet, compErr)
+		}
+	}
+	return nil
+}
+
+// Upload file
+func (b *Bucket) putFile(name string, r io.Reader) error {
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:212]: Comment not Meet godoc style.

File: pkg/objstore/ks3/ks3.go [hunk 3]
```diff
@@ -0,0 +1,521 @@
...
+					return errors.Wrapf(err, "callback func invoke for directory %s failed", *object.Prefix)
+				}
+			}
+			if !*objects.IsTruncated {
+				break
+			}
+			marker = *objects.NextMarker
+		}
+
+	} else {
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:267]: Return early will be better :)

File: pkg/objstore/ks3/ks3.go [hunk 4]
```diff
@@ -0,0 +1,521 @@
...
+			return nil, err
+		}
+		resp, err := b.ks3client.GetObject(&s3.GetObjectInput{
+			Bucket: &b.name,
+			Key:    &name,
+			Range:  &ranges,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("bucket:%v,object:%v,get range object failed,err:%v", b.name, name, err)
+		}
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:353]: Repeat Code.

File: pkg/objstore/ks3/ks3.go [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+		// bytes=-3-0
+		// bytes=-3--2
+		// are invalid.
+		return "", fmt.Errorf("set range failed")
+	}
+}
+
+// Exists checks if the given object exists.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	_, err := b.ks3client.GetObject(&s3.GetObjectInput{
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:386]: Should Response need Close?

File: pkg/objstore/ks3/ks3.go [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+		if b.IsObjNotFoundErr(err) {
+			return false, nil
+		}
+		return false, errors.Wrap(err, "ks3 getObject failed")
+	}
+	return true, nil
+}
+
+// IsObjNotFoundErr returns true if error means that object is not found. Relevant to Get operations.
+func (b *Bucket) IsObjNotFoundErr(err error) bool {
```
Comment (hanjm) [inline pkg/objstore/ks3/ks3.go:400]: Cloud it judge by type assert instead of `strings.Contains`?

**Review by hanjm — COMMENTED** [review]
Need update update `THANOS_TEST_OBJSTORE_SKIP` in file `.circleci/config.yml` `makefile`.

**Comment by hanjm** [comment]
💪🏻

The DCO check could done by `git commit --amend --signoff `.

**Comment by lizw91** [comment]
> Need update update `THANOS_TEST_OBJSTORE_SKIP` in file `.circleci/config.yml` `makefile`.

thanks

**Pull request closed without merging**